test/shared/src/main/scala/zio/test/TestClock.scala

```diff
@@ -260,131 +266,157 @@ object TestClock extends Serializable {
      * `TestClock` but a fiber is not suspending.
      */
     private[TestClock] def suspendedWarningDone(implicit trace: Trace): UIO[Unit] =
-      suspendedWarningState.updateSomeZIO[Any, Nothing] { case SuspendedWarningData.Pending(fiber) =>
-        fiber.interrupt.as(SuspendedWarningData.start)
+      suspendedWarningState.updateSomeZIO { case SuspendedWarningData.Pending(cancel) =>
+        ZIO.succeed {
+          cancel()
+          SuspendedWarningData.start
+        }
       }
 
     /**
      * Cancels the warning message that is displayed if a test is using time but
      * is not advancing the `TestClock`.
      */
     private[TestClock] def warningDone(implicit trace: Trace): UIO[Unit] =
-      warningState.updateSomeZIO[Any, Nothing] {
-        case WarningData.Start          => ZIO.succeed(WarningData.done)
-        case WarningData.Pending(fiber) => fiber.interrupt.as(WarningData.done)
+      warningState.updateSomeZIO {
+        case WarningData.Pending(cancelWarning) =>
+          ZIO.succeed {
+            cancelWarning()
+            WarningData.done
+          }
+        case _: WarningData.Start.type => Exit.succeed(WarningData.done)
       }
 
     /**
      * Polls until all descendants of this fiber are done or suspended.
      */
-    private def awaitSuspended(implicit trace: Trace): UIO[Unit] =
-      suspendedWarningStart *>
-        suspended
-          .zipWith(live.provide(ZIO.sleep(10.milliseconds)) *> suspended)(_ == _)
-          .filterOrFail(identity)(())
-          .eventually *>
-        suspendedWarningDone
+    private def awaitSuspended(waitFor: Duration)(implicit trace: Trace): UIO[Unit] =
+      freezeLock.withPermit(ZIO.suspendSucceed {
+        val ref = new AtomicBoolean(false)
+
+        def allSuspendedUnchanged(
+          l: collection.Map[FiberId, Fiber.Status.Suspended],
+          r: collection.Map[FiberId, Fiber.Status.Suspended]
+        ): Boolean = {
+          val it = r.iterator
+          while (it.hasNext) {
+            val rkv = it.next()
+            val lv  = l.getOrElse(rkv._1, null) // If missing, it wasn't previously created, so we have to return false
+            if (lv != rkv._2) return false
+          }
+          true
+        }
 
-    /**
-     * Delays for a short period of time.
-     */
-    private def delay(implicit trace: Trace): UIO[Unit] =
-      live.provide(ZIO.sleep(5.milliseconds))
+        // Sleep to give suspended fibers a chance to resume
+        val f = ClockLive.sleep(waitFor) *> freeze
+        f.zipWith(f)(allSuspendedUnchanged)
+          .flatMap {
+            if (_) Exit.boolean(ref.get)
+            else if (ref.compareAndSet(false, true)) suspendedWarningStart *> Exit.failUnit
+            else Exit.failUnit
+          }
+          .eventually
+          .flatMap(ZIO.whenDiscard(_)(suspendedWarningDone))
```

Comment:
this will involve `ZIO.suspendSucceed` and `.unit`


```suggestion
          .flatMap(if (_) suspendedWarningDone else Exit.unit)
```